Add blog article about cgroup v2 (#47342)
kubernetes/website: content/en/blog/_posts/2024-11-10-migrate-cgroup-v2.md

---
layout: blog
title: 'The Shift to cgroups v2 in Kubernetes: What You Need to Know'
date: 2024-11-10
slug: kubernetes-cgroups-v2-shift
author: >
   Paco Xu (DaoCloud)
---

> `cgroups` (control groups) are a Linux kernel feature used for managing system resources.

Reviewer: ```suggestion
In Linux, _cgroups_ (control groups) are a kernel feature used for managing system resources.
```

Kubernetes uses cgroups to allocate resources like CPU and memory to containers,
ensuring that applications run smoothly without interfering with each other.

> With the release of Kubernetes v1.31, cgroups v1 has been moved into [maintenance mode]/blog/2024/08/14/kubernetes-1-31-moving-cgroup-v1-support-maintenance-mode/).

Reviewer: ```suggestion
With the release of Kubernetes v1.31, support for v1 cgroup management has been moved into [maintenance mode](/blog/2024/08/14/kubernetes-1-31-moving-cgroup-v1-support-maintenance-mode/).
```

For cgroups v2, it graduated in v1.25 2 years ago.

Top FAQs are why we should migrate, what's the benifits and lost,
and what needs to be noticed when using cgroups v2.

> ## cgroups v1 problem, and solutions in cgroups v2

Reviewer 2: I think I would restructure this blog in the following manner:
```yaml
1. Introduction
2. Why Migrate to cgroup v2?
  a.  Benefits of cgroup v2
  b.  Limitations of cgroup v1 addressed by v2
3. Adopting cgroup v2
  a. Requirements
  b. Troubleshooting and Monitoring Tools
4. Conclusion
5. Further reading 
```

cgroups v1 and cgroups official doc can be found in

- [v1 doc](https://www.kernel.org/doc/Documentation/cgroup-v1/)
- [v2 doc](https://www.kernel.org/doc/Documentation/cgroup-v2.txt)

Let's enumerate some known issues.

### active_file memory is not considered as available memory

> There is [a known issue](/docs/concepts/scheduling-eviction/node-pressure-eviction/#active-file-memory-is-not-considered-as-available-memory) of page cache: [#43916](https://github.com/kubernetes/kubernetes/issues/43916).

Reviewer: (nit)
```suggestion
There is a [known issue](/docs/concepts/scheduling-eviction/node-pressure-eviction/#active-file-memory-is-not-considered-as-available-memory) around page cache: [#43916](https://github.com/kubernetes/kubernetes/issues/43916).
```

- In cgroups v1, we have no native solutions.
  Workarounds are setting larger memory limit for Pods or using some external projects to drop cache or
  throttling memory allocating when memory is beyond a threshold.
- In cgroups v2, we can use `memory.high` to throttle.

Support for Memory QoS was initially added in Kubernetes v1.22, 
and later some limitations around the formula for calculating `memory.high` were identified.

> These limitations are addressed in Kubernetes v1.27.

Reviewer: ```suggestion
These limitations were addressed in Kubernetes v1.27.
```

> However, until v1.31, the feature gate is still alpha due to another known issue
> that application pod may be hanging forever due to heavy memory reclaiming.

Reviewer: ```suggestion
However, until v1.31, the feature gate for memory QoS was still alpha due to another known issue
(that an application pod may be hanging forever due to heavy memory reclaiming).
```

   - Cilium will automatically mount cgroups v2 filesystem required to attach BPF cgroup programs
     by default at the path /run/cilium/cgroupv2 .
2. PSI is planned in a future release [KEP-4205](https://github.com/kubernetes/enhancements/issues/4205),
   but pending due to runc 1.2.0 release delay.
3. monitoring tools support, like [Cadvisor](https://github.com/google/cadvisor/).
   Currently, cgroups v2 features are not fully-supported yet.

## Adopting cgroup version 2

### Requirements

Here's what you need to use cgroup v2 with Kubernetes.
First up, you need to be using a version of Kubernetes with support for v2 cgroup management;
that's been stable since Kubernetes v1.25 and all supported Kubernetes releases include this support.

- OS distribution enables cgroups v2
- Linux Kernel version is 5.8 or later
- Container runtime supports cgroups v2. For example:
  - containerd v1.4 or later (at the time of writing, containerd releases v1.6 and later are within that project's support period)
  - CRI-O v1.20 or later
- The kubelet and the container runtime are configured to use the systemd cgroup driver

#### kernel updates around cgroups v2

> cgroups v2 first appeared in Linux Kernel 4.5 in 2016.

Reviewer: How about:
```suggestion
When Kubernetes was first announced, in 2014, only v1 cgroups existed.
Version 2 cgroup management first appeared in Linux kernel 4.5, released in 2016.
```

- The Kubernetes project does not recommend using cgroups v2 with a Linux kernel older than 5.2 due to lack of cgroup-level task freezer support.
- In Kubernetes, 5.8 is the minimal kernel version for cgroups v2 as root `cpu.stat` file on cgroupv2
  was only added on kernel 5.8.
- `memory.peak` is added in 5.19.

### Use systemd as cgroup driver

[Configure the kubelet's cgroup driver to match the container runtime cgroup driver](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver/).

The [Container runtimes](/docs/setup/production-environment/container-runtimes)
page explains that the `systemd` driver is recommended for kubeadm
based setups instead of the kubelet's default `cgroupfs` driver,
because kubeadm manages the kubelet as a systemd service.

A minimal example of configuring the field explicitly:

```yaml
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
cgroupDriver: systemd
```

In v1.31, [KEP-4033](https://github.com/kubernetes/enhancements/issues/4033) is beta to extend CRI API for the kubelet
to discover the cgroup driver from the container runtime. This will help installer and kubelet to autodetect

> - TODO

Reviewer: This doesn't look ready for review.

### Tools and commands for troubleshooting

Tools and commands that you should know about cgroups:

- `stat -fc %T /sys/fs/cgroup/`: Check if cgroups v2 is enabled which will return `cgroup2fs`
- `systemctl list-units kube* --type=slice` or `--type=scope`: List kube related units that systemd currently has in memory.

> - `bpftool cgroup list /sys/fs/cgroup/*`: List all programs attached to the cgroup CGROUP.

Reviewer 3: `bpftool cgroup list /sys/fs/cgroup/*` actually shows all programs?
In my environment, I could not list it.
First of all, I think `bpftool cgroup tree` is suitable to list all  attached  programs.

BTW, I have recently fixed the bug of `bpftool cgroup` as below:
https://patchwork.kernel.org/project/netdevbpf/patch/20240607111704.6716-1-tadakentaso@gmail.com/
So if you want to try bpftool cgroup tree` in your environment, please keep that in your mind.

---
layout: blog

> title: 'Why we should migrate to vgroup v2?'

Reviewer: Early nit:
```suggestion
title: 'Why we should migrate to cgroup v2?'
```

Author: Changed a title: `The Shift to cgroup v2 in Kubernetes: What You Need to Know`.

Because this blog doesn't focus on how to migrate but why, FAQ and what users should notice.
